### ISSUE TITLE
JavadocMethod to have dedicated switch for type param

### ISSUE THREAD

**Issue by ladislavlencucha** [checkstyle/checkstyle#10003]
Currently, JavadocMethod supports switches/properties to allow missing param tag and missing return tag:
https://checkstyle.sourceforge.io/config_javadoc.html#JavadocMethod

I'd like to be able to enforce all method parameters to be present in javadoc (fail if missing), but not the generic types (allow missing type param). Therefore, I propose to add one additional property _allowMissingTypeParamTags_ which is true by default. It might cause upgrade issues for those using allowMissingParamTags, but is cleaner than having some switch applied when other switch is true.

I can provide pull request with implementation, if you decide to include this feature and that this is the correct way.

/var/tmp $ cat config.xml
<module name="TreeWalker">
  <module name="JavadocMethod">
    <property name="allowMissingTypeParamTags" value="true"/>
  </module>
</module>

/var/tmp $ javac Test.java

/var/tmp $ cat Test.java
public interface Test {
  /**
    * This method will no be reported due to missing param <T>
    * @param variable
    * @return bla bla
    */
  <T> T methodWithCorrectJavadoc(T variable);  // To be checked

  /**
    * Report this method due to missing param <T>
    * @return bla bla
    */
  <T> T methodWithIncorrectJavadoc(T variable);  // To be checked
}

**Comment by nmancus1** [comment]
These violations could be suppressed using [SuppressionSingleFilter](https://checkstyle.sourceforge.io/config_filters.html#SuppressionSingleFilter):

```bash

➜  src cat Test.java 
    public interface Test {
        /**
         * This method will not be reported due to missing param <T>
         * @param variable
         * @return bla bla
         */
        <T> T methodWithCorrectJavadoc(T variable);  // ok

        /**
         * Report this method due to missing param variable
         * @return bla bla
         */
        <T> T methodWithIncorrectJavadoc(T variable);  // violation
}

➜  src cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="SuppressionSingleFilter">
        <property name="message" value="Expected @param tag for '&lt;"/>
    </module>

    <module name="TreeWalker">
        <module name="JavadocMethod"/>

    </module>
</module>

➜  src java -jar checkstyle-8.42-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/IdeaProjects/Checkstyle-test/src/Test.java:13:44: Expected @param tag for 'variable'. [JavadocMethod]
Audit done.
Checkstyle ends with 1 errors.

```

In the future, we will have a better way to suppress by message: https://github.com/checkstyle/checkstyle/issues/3082

**Comment by ladislavlencucha** [comment]
Nice workaround, thanks! Anyway, I find my proposal more consistent (those 2 existing properties could have been implemented exactly the same way as you propose).

Anyway, if you think it is of no use I accept. If you think it could be useful I am more than willing to contribute.

Thanks again!

**Comment by nmancus1** [comment]
We will see what members have to say, but I know that we have been favoring the suppression/ filter model over adding additional properties to checks, especially those that already have a lot of properties (such as JavadocMethod), especially when a particular suppression is straightforward and reliable, i.e.would not cause false positives/ negatives.

**Comment by ladislavlencucha** [comment]
Makes sense and is actually more flexible. On the other hand, it is also fragile in case somebody changes e.g. the message.
Thanks for explaining what the aim is and your time :-)

**Comment by romani** [comment]
I can not approve this issue, as we are not ready to continue extend Check that is based on Regexp to parse.
I am keep this issue open as reminder of user interest of behavior customization. Very reasonable demand.
We will comeback to it ones we start split of such Check to more specific Checks with parsing base on grammar.

**Comment by AashishShrestha97** [comment]
Hi @ladislavenuca @romani and Checkstyle team,

I’d like to work on this issue (#10003) to add a new `allowMissingTypeParamTags` property to `JavadocMethodCheck` with a default value of `true`, as proposed. The idea is to separately control violations for missing type parameter `@param` tags while keeping `allowMissingParamTags` for regular parameters. I plan to:
- Add the new field and setter in `JavadocMethodCheck.java`.
- Update the class Javadoc and split the violation check logic.
- Add corresponding unit tests and update the documentation.

Please let me know if this approach works or if there are specific requirements, I should consider before proceeding with a PR. Looking forward to your feedback!

Thanks,
AashishShrestha97

**Comment by romani** [comment]
@AashishShrestha97 , issue is not approved, it means , any fix will not merged. You can copy whole code of Check to your own project (extension of checkstyle) to make any changes you like.

Users should use suggested suppression to avoid such violations.

**Comment by romani** [comment]
suppression should be used 
https://github.com/checkstyle/checkstyle/issues/10003#issuecomment-840554396

if user do not like something in particular from Check behavior, suppression is way to reconcile requirements.
